### ISSUE TITLE
Upgrade 3.3.5 -> 3.4.5

### ISSUE THREAD

**Issue by Quadrocompile** [webanno/webanno#1022]
We experienced some server issues which led to wiping the server and reinstalling webanno 3.4.5 from scratch. We have a copy of the old webanno home ( /srv/webanno and /opt/webanno ) as well as a complete backup of the old database (and a separate one with only the users and authorities tables). The old files are from the 3.3.5 version.

After installing WebAnno 3.4.5 we placed the old contents of /srv/webanno in the new directory and changed the user to www-data. We also imported the users and authorities tables into mysql. We are now able to log into WebAnno as admin and the users are restored correctly, **however, none of the projects appears in the project list**. I do not know if the other users could access their projects, but logged in as admin user it states _You are not member of any projects to annotate or curate_ at the bottom of the screen. The backup files size is about 1GB of data, so I assume that the backup is complete. Is WebAnno 3.4.5 incompatible with project files created with WebAnno 3.3.5 or is there something we missed when restoring the files?

Kind regards

**Comment by reckart** [comment]
The complete database needs to be available to the upgraded WebAnno, not only the users and authorities tables. All metadata about projects, annotation states, etc. are stored in the database.

**Comment by Quadrocompile** [comment]
fair enough, i stopped the webanno service, imported the backup of the whole database but then the service won't startup again -> **ERROR [SYSTEM] SpringApplication - Application startup failed**
Were there changes to the database scheme between v3.3.5 and 3.4.5 and is it possible to update the old database to the new version? One line in the stack trace states **Schema-validation: missing column [traits] in table [annotation_feature]** maybe this is helpful.

Attached the corresponding catalina log file and the used database dump from webanno 3.3.5. 
[webannodb_complete.zip](https://github.com/webanno/webanno/files/2229482/webannodb_complete.zip)

[catalina.out.txt](https://github.com/webanno/webanno/files/2229480/catalina.out.txt)

**Comment by Quadrocompile** [comment]
Would it help to deploy webanno version 3.4.0 first to allow the 3.3.5 database to be upgraded to the new 3.4.x datamodel before using 3.4.5 again?

**Comment by reckart** [comment]
When coming from any older version, an upgrade to 3.3.5 is highly mandatory before upgrading to a 3.4.x version. But an intermediate upgrade to 3.4.0 before going higher should not be necessary.

WebAnno now uses Liquibase to manage the database schema. During startup, Liquibase checks the status of the database and applies any change sets that haven't been applied yet. After that, Hibernate kicks in to validate if the schema created by Liquibase matches its expectations.

In your current DB (not in the `webannodb_complete.zip` dump that you sent), there should be a table called `DATABASECHANGELOG` and in that table, there should be an entry with the ID `20180331-1`. Can you please check that table exists and that it is marked as "executed"?

And please ensure that WebAnno as full permissions on its database and all of the tables in the database.

**Comment by reckart** [comment]
Btw: I would strongly recommend to delete your comment containing the DB dump - it contains usernames, emails, encrypted passwords - I kind of doubt you want to keep this floating around on the net.

**Comment by Quadrocompile** [comment]
```console
select * from DATABASECHANGELOG WHERE ID = "20180331-1";
+------------+--------------+----------------------------------------------------------+---------------------+---------------+----------+------------------------------------+----------------------------------------+----------+------+-----------+----------+--------+---------------+
| ID         | AUTHOR       | FILENAME                                                 | DATEEXECUTED        | ORDEREXECUTED | EXECTYPE | MD5SUM                             | DESCRIPTION                            | COMMENTS | TAG  | LIQUIBASE | CONTEXTS | LABELS | DEPLOYMENT_ID |
+------------+--------------+----------------------------------------------------------+---------------------+---------------+----------+------------------------------------+----------------------------------------+----------+------+-----------+----------+--------+---------------+
| 20180331-1 | WebAnno Team | de/tudarmstadt/ukp/clarin/webanno/model/db-changelog.xml | 2018-07-25 19:06:42 |            78 | EXECUTED | 7:6f94d2fd825dd714eafb9198debea53f | addColumn tableName=annotation_feature |          | NULL | 3.5.3     | NULL     | NULL   | 2538394387    |
+------------+--------------+----------------------------------------------------------+---------------------+---------------+----------+------------------------------------+----------------------------------------+----------+------+-----------+----------+--------+---------------+
1 row in set (0.00 sec)
```
Strange, this suggests that the missing entry has been created, right? When exactly will Liquibase attempt to convert the database scheme?

**Comment by reckart** [comment]
Liquibase kicks is very early during the startup of the spring context before Hibernate/JPA is started.

Can you post the current table layout of the `annotation_feature` table? I assume that "traits" must be there if the change set has executed.

**Comment by reckart** [comment]
@Quadrocompile Happy it worked out :) Unfortunately, we recently ran into a dead-end relying only on Hibernate managing the DB scheme which forced the switch to Liquibase. But honestly, it's really the better approach as it now allows properly refactoring the DB as part of upgrades and addressing several long-standing issues in the database schema. Relying just on Hibernate was really living on borrowed time - was ok for some years, but now Liquibase is here to stay.
